### PR TITLE
MOD-5740: Default querying language

### PR DESCRIPTION
**Pull request by nafraf**
**Description**

1. When the language is not specified in the query, the engine should be smart enough to follow the index defined language, instead of having wrong results or no results.
2. The query should follow the LANGUAGE specified on the Global index (not the default).

**Which issues this PR fixes**
1. MOD-5740

**Main objects this PR modified**
1. ...
3. ...

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: tests/pytests/test_language.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    # Create sample data in Italian
+    conn.execute_command('HSET', '{word}:1', 'word', 'arancia') # orange
+    conn.execute_command('HSET', '{word}:2', 'word', 'arance') # oranges
+
+    # Create sample data in English
+    conn.execute_command('HSET', '{word}:3', 'word', 'orange') # orange
+    conn.execute_command('HSET', '{word}:4', 'word', 'oranges') # oranges
+
+    # Wait for index to be created
+    waitForIndex(env, 'idx')
```
Comment (raz-mon) [inline tests/pytests/test_language.py:22]: `HSET` is synchronous, so no need for the wait

File: tests/pytests/test_language.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    expected = [2, '{word}:3', ['word', 'orange'], '{word}:4', ['word', 'oranges']]
+    env.assertEqual(res, expected)
+
+    # Search for "orange" in Italian
+    # should return 1 results using invalid stemming in Italian
+    res = env.cmd('FT.SEARCH', 'idx', 'oranges', 'language', 'italian')
+    env.assertEqual(res, [1, '{word}:4', ['word', 'oranges']])
+
+    # drop index
+    env.cmd('FT.DROP', 'idx')
```
Comment (raz-mon) [inline tests/pytests/test_language.py:48]: Why is this needed?

File: tests/pytests/test_language.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    # Create sample data in Italian
+    conn.execute_command('JSON.SET', '{word}:1', '$', '{"word":"arancia"}') # orange
+    conn.execute_command('JSON.SET', '{word}:2', '$', '{"word":"arance"}') # oranges
+
+    # Create sample data in English
+    conn.execute_command('JSON.SET', '{word}:3', '$', '{"word":"orange"}') # orange
+    conn.execute_command('JSON.SET', '{word}:4', '$', '{"word":"oranges"}') # oranges
+
+    # Wait for index to be created
+    waitForIndex(env, 'idx')
```
Comment (raz-mon) [inline tests/pytests/test_language.py:67]: `HSET` is synchronous, so no need for the wait

File: tests/pytests/test_language.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+                  'SORTBY', 'word', 'ASC', 'NOCONTENT')
+    env.assertEqual(res, [2, '{word}:3', '{word}:4'])
+
+    # Search for "orange" using Italian
+    # It should return 1 results using invalid stemming in Italian
+    res = env.cmd('FT.SEARCH', 'idx', 'oranges', 'language', 'italian')
+    env.assertEqual(res, [1, '{word}:4', ['$', '{"word":"oranges"}']])
+
+    # drop index
+    env.cmd('FT.DROP', 'idx')
```
Comment (raz-mon) [inline tests/pytests/test_language.py:92]: Why is this needed? Please remove if not

File: tests/pytests/test_language.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+                         '__lang', 'italian') # apples
+    conn.execute_command('HSET', '{word}:5', 'word', 'orange',
+                         '__lang', 'english')
+    conn.execute_command('HSET', '{word}:6', 'word', 'oranges',
+                         '__lang', 'english')
+
+    # Create index - language Italian by default
+    env.cmd('FT.CREATE', 'idx', 'ON', 'HASH', 'PREFIX', '1', '{word}:',
+            'LANGUAGE', 'italian', 'LANGUAGE_FIELD', '__lang',
+            'SCHEMA', 'word', 'TEXT', '__lang', 'TEXT')
```
Comment (raz-mon) [inline tests/pytests/test_language.py:114]: Why do you need the last two argument here (`__lang`, `TEXT`)?

File: tests/pytests/test_language.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+    # Wait for index to be created
+    waitForIndex(env, 'idx')
```
Comment (raz-mon) [inline tests/pytests/test_language.py:117]: I think this isn't doing anything here.. Please remove if so.

File: tests/pytests/test_language.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    res = env.cmd('FT.SEARCH', 'idx', 'arancia', 'SORTBY', 'word', 'DESC', 'LANGUAGE', 'english')
+    expected = [1, '{word}:1', ['word', 'arancia']]
+    env.assertEqual(res, expected)
+
+    # Search for "oranges", passing language argument will override the language in the schema
+    # It should return 2 results using stemming in English
+    res = env.cmd('FT.SEARCH', 'idx', 'oranges', 'SORTBY', 'word', 'ASC', 'LANGUAGE', 'english')
+    expected = [2, '{word}:5', ['word', 'orange'], '{word}:6', ['word', 'oranges']]
+    env.assertEqual(res, expected)
+
```
Comment (raz-mon) [inline tests/pytests/test_language.py:144]: How about adding a test searching for "orange" without a language in the query? Just to see that the default language of the index is used, and you get one result (orange..).

File: src/aggregate/aggregate_request.c [hunk 8]
```diff
@@ -568,7 +568,13 @@ static int parseQueryArgs(ArgsCursor *ac, AREQ *req, RSSearchOptions *searchOpts
   searchOpts->ninkeys = inKeys.argc;
   searchOpts->legacy.infields = (const char **)inFields.objs;
   searchOpts->legacy.ninfields = inFields.argc;
-  searchOpts->language = RSLanguage_Find(languageStr, 0);
+  // if language is NULL, set it to UNDEFINED and it will be updated later,
+  // taking the index language
+  if(languageStr == NULL) {
+    searchOpts->language = RS_LANG_INDEX_DEFAULT;
```
Comment (raz-mon) [inline src/aggregate/aggregate_request.c:574]: The doc say `UNDEFINED`, but the code says `DEFAULT`.

File: src/aggregate/aggregate_request.c [hunk 9]
```diff
@@ -986,10 +992,13 @@ int AREQ_ApplyContext(AREQ *req, RedisSearchCtx *sctx, QueryError *status) {
     }
   }
 
-  if (opts->language == RS_LANG_UNSUPPORTED) {
+  if (opts->language == RS_LANG_INDEX_DEFAULT) {
+    opts->language = index->rule->lang_default;
```
Comment (raz-mon) [inline src/aggregate/aggregate_request.c:996]: What is the default initialized to?
If it's not set by the user, is it English?

**Review by raz-mon — COMMENTED** [review]
Nice 👍 
See some comments

File: tests/pytests/test_language.py (hunk 2 above)
Comment (nafraf) [reply tests/pytests/test_language.py]: It is not needed, I removed it.

File: tests/pytests/test_language.py (hunk 4 above)
Comment (nafraf) [reply tests/pytests/test_language.py]: I removed it.

File: tests/pytests/test_language.py (hunk 5 above)
Comment (nafraf) [reply tests/pytests/test_language.py]: You are right, I don'n need it. I removed it.

File: src/aggregate/aggregate_request.c (hunk 8 above)
Comment (nafraf) [reply src/aggregate/aggregate_request.c]: OK, I updated the comment. Thanks

File: tests/pytests/test_language.py (hunk 7 above)
Comment (nafraf) [reply tests/pytests/test_language.py:159]: I added more tests. The last one is still failing. I need to fix it.

File: src/aggregate/aggregate_request.c (hunk 9 above)
Comment (nafraf) [reply src/aggregate/aggregate_request.c:996]: Yes, if it is not set by the user, the default language is English

File: src/language.h [hunk 10]
```diff
@@ -4,52 +4,53 @@
...
+  RS_LANG_RUSSIAN,
+  RS_LANG_SPANISH,
+  RS_LANG_SWEDISH,
+  RS_LANG_TAMIL,
+  RS_LANG_TURKISH,
+  RS_LANG_ARMENIAN,
+  RS_LANG_SERBIAN,
+  RS_LANG_YIDDISH,
+  RS_LANG_UNSUPPORTED,
+  RS_LANG_INDEX_DEFAULT, // The user did not set the language for FT.SEARCH, use the index language
```
Comment (oshadmi) [inline src/language.h:45]: Consider a neutral value such as `RS_LANG_UNSET` etc., that will keep the logic outside of the enum.

File: src/aggregate/aggregate_request.c [hunk 11]
```diff
@@ -568,7 +568,13 @@ static int parseQueryArgs(ArgsCursor *ac, AREQ *req, RSSearchOptions *searchOpts
   searchOpts->ninkeys = inKeys.argc;
   searchOpts->legacy.infields = (const char **)inFields.objs;
   searchOpts->legacy.ninfields = inFields.argc;
-  searchOpts->language = RSLanguage_Find(languageStr, 0);
+  // if language is NULL, set it to RS_LANG_INDEX_DEFAULT and it will be updated
+  // later, taking the index language
+  if(languageStr == NULL) {
+    searchOpts->language = RS_LANG_INDEX_DEFAULT;
+  } else {
+    searchOpts->language = RSLanguage_Find(languageStr, 0);
```
Comment (oshadmi) [inline src/aggregate/aggregate_request.c:576]: Can `RSLanguage_Find` return `RS_LANG_UNSET` on NULL?

What about LLAPI (low-level API) such as `RediSearch_CreateDocument`,
Need to also add the same logic there?

File: src/language.c [hunk 12]
```diff
@@ -66,7 +66,7 @@ const char *RSLanguage_ToString(RSLanguage language) {
     case  RS_LANG_IRISH:       ret = "irish";      break;
     case  RS_LANG_ITALIAN:     ret = "italian";    break;
     case  RS_LANG_LITHUANIAN:  ret = "lithuanian"; break;
-    case  RS_LANG_NEPALI:      ret = "napali";     break;
+    case  RS_LANG_NEPALI:      ret = "nepali";     break;
```
Comment (oshadmi) [inline src/language.c:69]: Can add a test (using `FT.INFO`?)

File: tests/pytests/test_json.py [hunk 13]
```diff
@@ -718,17 +718,17 @@ def testNumeric(env):
     env.expect('FT.SEARCH', 'idx', '@f:[9.5 9.9]', 'RETURN', '3', '$.f', 'AS', 'flt') \
         .equal([1, 'doc:1', ['flt', '9.72']])
 
-@skip()
+@skip(cluster=True)
```
Comment (oshadmi) [inline tests/pytests/test_json.py:721]: Can also be tested on cluster?

**Review by oshadmi — COMMENTED** [review]
👍🏼 Some comments

File: tests/pytests/test_language.py [hunk 14]
```diff
@@ -0,0 +1,283 @@
...
+    # To validate this, we check that we have the same results in both cases:
+    # 1. passing the language argument = English
+    # 2. searching without language argument
+    for words in ['arancia', 'arance', 'ciliegia', 'ciliegie']:
+        res1 = env.cmd('FT.SEARCH', 'idx_en', 'arance', 'SORTBY', 'word', 'DESC')
+        res2 = env.cmd('FT.SEARCH', 'idx_en', 'arance', 'SORTBY', 'word', 'DESC',
+                        'LANGUAGE', 'english')
+        env.assertEqual(res1, res2)
+
+
```
Comment (oshadmi) [inline tests/pytests/test_language.py:283]: Consider testing using the lang field such as 
`@lang:(italian)  ciliegia LANGUAGE ITALIAN`
Or
`@lang:(italian)  ciliegia` with index language `Italian` ?

File: src/language.h (hunk 10 above)
Comment (oshadmi) [reply src/language.h:45]: Need to double check the new value doesn't get to `SchemaRule_RdbSave`

File: tests/pytests/test_json.py (hunk 13 above)
Comment (nafraf) [reply tests/pytests/test_json.py]: Yes. Now the test run on cluster.

File: tests/pytests/test_language.py (hunk 14 above)
Comment (nafraf) [reply tests/pytests/test_language.py]: New tests included in [2b3dbc2](https://github.com/RediSearch/RediSearch/pull/4455/commits/2b3dbc24fbde982dde73e1de38282cb88e497f21)

Comment (nafraf) [reply tests/pytests/test_language.py]: Tests included in [2b3dbc2](https://github.com/RediSearch/RediSearch/pull/4455/commits/2b3dbc24fbde982dde73e1de38282cb88e497f21)

File: src/aggregate/aggregate_request.c (hunk 11 above)
Comment (nafraf) [reply src/aggregate/aggregate_request.c:576]: No, `RS_LANG_UNSET` is used only to identify that the user did not pass any language in `FT.QUERY` command but the logic of `RSLanguage_Find` was not modified, and should not return `RS_LANG_UNSET.
As `RSLanguage_Find` was not modified, I don't see any problem with the LLAPI or need to change it.

File: src/language.h (hunk 10 above)
Comment (nafraf) [reply src/language.h]: > Consider a neutral value such as RS_LANG_UNSET etc., that will keep the logic outside of the enum.

A neutral value is now used: `RS_LANG_UNSET`

> Need to double check the new value doesn't get to `SchemaRule_RdbSave`

These are the validations to check that the value `RS_LANG_UNSET` does not get to `SchemaRule_RdbSave`:
1. The value `RS_LANG_UNSET` is only assigned to the variable used for the search `RSSearchOptions->language`, it is not assigned to schema rule variable
2. The schema rule variable ``SchemaRule->lang_default` receives the result of the function `RSLanguage_Find()`, but that functions does not return `RS_LANG_UNSET`
3. `RSSearchOptions->language` is not assigned to `SchemaRule->lang_default`

File: tests/pytests/test_language.py (hunk 1 above)
Comment (nafraf) [reply tests/pytests/test_language.py]: Thanks. The `waitForIndex()` is now called after the `FT.CREATE`, in that case it is required to avoid a flaky test.

File: src/aggregate/aggregate_request.c (hunk 11 above)
Comment (nafraf) [reply src/aggregate/aggregate_request.c:581]: LLAPI is working, the change was done in `redisearch_api.c`, if the index contains a language different from default, it is assigned to `RSSearchOptions->language`.
Two tests were created:
* `TEST_F(LLApiTest, testIndexWithDefaultLanguage)`
* `TEST_F(LLApiTest, testIndexWithCustomLanguage)`

**Review by oshadmi — APPROVED** [review]
👍🏼 
Do we need a completion task for adding LLAPI to accept a specific LANG in a query?

**Review by raz-mon — DISMISSED** [review]
💪

**Comment by nafraf** [comment]
> 👍🏼 Do we need a completion task for adding LLAPI to accept a specific LANG in a query?

Yes. MOD-6871 was created.

**Review by oshadmi — APPROVED** [review]
👍🏼

**Review by oshadmi — APPROVED** [review]
